xain_fl/grpc/participant.py

```diff
@@ -0,0 +1,377 @@
+            # which one was it?
+            return self.state
+
+    def wait_until_next_round(self):
+        """Waits until Participant is in a state indicating the start of the next round
+        of training.
+
+        Returns:
+            :obj:`ParState`: New state Participant is in.
+        """
+        with self.cv:
+            self.cv.wait_for(
+                lambda: self.state
+                in {ParState.TRAINING, ParState.WAITING_FOR_SELECTION, ParState.DONE}
+            )
+            # which one was it?
+            return self.state
+
+
+def transit(st, beat_reply):
+    """Participant state transition function on a heartbeat response. Updates the
+    state record `st`.
+
+    Args:
+        st (obj:`StateRecord`): Participant state record to update.
+        beat_reply (obj:`coordinator_pb2.HeartbeatReply`): Heartbeat from Coordinator.
+
+    """
+    msg, r = beat_reply.state, beat_reply.round
+    with st.cv:
+        if st.state == ParState.WAITING_FOR_SELECTION:
+            if msg == coordinator_pb2.State.Value("ROUND"):
+                st.state = ParState.TRAINING
+                st.round = r
+                st.cv.notify()
+            elif msg == coordinator_pb2.State.Value("FINISHED"):
+                st.state = ParState.DONE
+                st.cv.notify()
+        elif st.state == ParState.POST_TRAINING:
+            if msg == coordinator_pb2.State.Value("STANDBY"):
+                # not selected
+                st.state = ParState.WAITING_FOR_SELECTION
+                # prob ok to keep st.round as it is
+                st.cv.notify()
+            elif msg == coordinator_pb2.State.Value("ROUND") and r == st.round + 1:
+                st.state = ParState.TRAINING
+                st.round = r
+                st.cv.notify()
+            elif msg == coordinator_pb2.State.Value("FINISHED"):
+                st.state = ParState.DONE
+                st.cv.notify()
+
+
+def message_loop(chan, st, terminate):
+    """Periodically sends (and handles) heartbeat messages in a loop.
+
+    Args:
+        chan: gRPC channel to Coordinator.
+        st (obj:`StateRecord`): Participant state record.
+        terminate (obj:`threading.Event`): Event to terminate message loop.
+    """
+    coord = coordinator_pb2_grpc.CoordinatorStub(chan)
+    while not terminate.is_set():
+        start = datetime.today()
+        print(f"sending heartbeat {start}")
+        if st.state in [ParState.WAITING_FOR_SELECTION, ParState.POST_TRAINING, ParState.DONE]:
+            state = coordinator_pb2.State.Value("READY")
+        else:
+            state = coordinator_pb2.State.Value("TRAINING")
+        req = coordinator_pb2.HeartbeatRequest(state=state, round=st.round)
+        reply = coord.Heartbeat(req)
+        transit(st, reply)
+        end = datetime.today() - start
+        print(f"ended sending heartbeat {end}")
+        time.sleep(HEARTBEAT_TIME)
+
+
+def go(part: Participant, coordinator_address: str):
```

Comment:
Add the ability to pass in a coordinator address instead of using an harcoded one.